packages/eslint-plugin/tests/rules/no-misused-promises.test.ts

```diff
@@ -970,5 +970,29 @@ console.log({ ...(condition ? Promise.resolve({ key: 42 }) : {}) });
         { line: 7, messageId: 'spread' },
       ],
     },
+    {
+      code: `
+type MyUnion = (() => void) | boolean;
+
+function restPromises(first: Boolean, ...callbacks: Array<() => void>): void {}
+function restUnion(first: string, ...callbacks: Array<MyUnion>): void {}
+
+restPromises(
+  true,
+  () => Promise.resolve(true),
+  () => Promise.resolve(null),
+  () => true,
+  () => Promise.resolve('Hello'),
+);
+
+restUnion('Testing', false, () => Promise.resolve(true));
+      `,
+      errors: [
+        { line: 9, messageId: 'voidReturnArgument' },
+        { line: 10, messageId: 'voidReturnArgument' },
+        { line: 12, messageId: 'voidReturnArgument' },
+        { line: 15, messageId: 'voidReturnArgument' },
+      ],
+    },
```

Comment:
These are very good tests to start, but I think we'll want a few more for more edge cases around tuples: `[]`, `[() => void]`, `[undefined, () => void]`, etc.

There should also be some `valid` cases added too, to make sure the rule isn't giving false reports. `Array<() => Promise>` comes to mind there.

Reply:
I've added some valid tests (as well as a valid test for `new MyClass(() => Promise.resolve(1))`)